odoo/fields.py

```diff
@@ -577,12 +581,18 @@ def _compute_related(self, records):
         """ Compute the related field ``self`` on ``records``. """
         # when related_sudo, bypass access rights checks when reading values
         others = records.sudo() if self.related_sudo else records
-        for record, other in zip(records, others):
-            if not record.id and record.env != other.env:
-                # draft records: copy record's cache to other's cache first
-                copy_cache(record, other.env)
-            other, field = self.traverse_related(other)
-            record[self.name] = other[field.name]
+        # copy new records' cache into others' cache
+        if records.env != others.env:
+            copy_cache(records - records.filtered('id'), others.env)
+        # traverse fields one by one for all records, in order to take advantage
+        # of prefetching at each level. This is important specially for the last
+        # field, in case it is a computed field!
+        values = list(others)
+        for name in self.related[:-1]:
+            values = [first(value[name]) for value in values]
```

Comment:
Am I wrong by saying that this change should also reduce the number of SQL queries to get the related fields for records in draft mode?

Reply:
@lmignon it is a consequence of the point above: computing fields in batch implies that the records being computed "populate" the prefetching and therefore generate less queries to read the appropriate data.